core/lens/lens_phrase.ml

```diff
@@ -0,0 +1,235 @@
+open Utility
+open Types
+open Sugartypes
+open Lens_operators
+open Lens_utility
+
+module Alias = Lens_alias
+
+type t = lens_phrase
+
+let var n = Var n
+
+let and' v1 v2 = InfixAppl (Binary.Logical Logical_binop.And, v1, v2)
+
+let or' v1 v2 = InfixAppl (Binary.Logical Logical_binop.Or, v1, v2)
+
+let equal v1 v2 = InfixAppl (Binary.Equal, v1, v2)
+
+let not' v1 = UnaryAppl (Unary.Not, v1)
+
+let tuple v = TupleLit v
+
+let tuple_singleton v = tuple [v]
+
+let name_of_var expr =
+  match expr.node with
+  | `Var n -> n
+  | _ -> failwith "Expected var."
+
+let of_phrase p =
+  let rec f p =
+    match p.node with
+    | `Constant c -> Constant c
+    | `Var v -> Var v
+    | `UnaryAppl ((_, op), phrase) -> UnaryAppl (Unary.from_links op, f phrase)
+    | `InfixAppl ((_, op), phrase1, phrase2) -> InfixAppl (Binary.from_links op, f phrase1, f phrase2)
+    | `TupleLit l -> TupleLit (List.map f l)
+    | `FnAppl (fn, arg) ->
+      begin
+        match name_of_var fn with
+        | "not" -> UnaryAppl ((Unary.Name "!"), f (List.hd arg))
+        | _ -> failwith "Unsupported function"
+      end
+    | _ -> failwith "Unknown phrasenode for lens_phrase to phrase."
+  in
+  f p
+
+let rec traverse (expr : lens_phrase) dosth =
```

Comment:
What is `dosth`? Is it an abbreviation for `do something`? I suggest naming it just `f`, and possibly changing the order of arguments, then a functional programmer would immediately recognise this as map-like function. Or if you prefer a more descriptive, then I suggest you use `action`.

Reply:
Changed.